### ISSUE TITLE
mutex not m_initialised during unmount

### ISSUE THREAD

**Issue by EchterAgo** [openzfsonwindows/openzfs#303]
I create this issue to continue discussion of the issue described in https://github.com/openzfsonwindows/openzfs/issues/283#issuecomment-1772518302

---

With cab2a207eccf2666e77c981de345f8bcf3b3125c I get a `XÎ›#øÿÿ: mutex not m_initialised`

```
 # Child-SP          RetAddr               Call Site
00 ffff870e`668845f0 fffff802`23497fec     OpenZFS!panic+0x3c [H:\dev\openzfs\module\os\windows\spl\spl-debug.c @ 32] 
01 ffff870e`66884630 fffff802`23498ec8     OpenZFS!spl_mutex_enter+0x3c [H:\dev\openzfs\module\os\windows\spl\spl-mutex.c @ 119] 
02 ffff870e`668846a0 fffff802`23534e7d     OpenZFS!spl_cv_wait+0xf8 [H:\dev\openzfs\module\os\windows\spl\spl-condvar.c @ 127] 
03 ffff870e`66884730 fffff802`23534d1f     OpenZFS!rrw_enter_read_impl+0x14d [H:\dev\openzfs\module\zfs\rrwlock.c @ 178] 
04 ffff870e`66884780 fffff802`23535709     OpenZFS!rrw_enter_read+0x1f [H:\dev\openzfs\module\zfs\rrwlock.c @ 198] 
05 ffff870e`668847c0 fffff802`2377d2e4     OpenZFS!rrm_enter_read+0x49 [H:\dev\openzfs\module\zfs\rrwlock.c @ 364] 
06 ffff870e`66884800 fffff802`2377d1be     OpenZFS!zfs_enter+0x24 [H:\dev\openzfs\include\os\windows\zfs\sys\zfs_znode_impl.h @ 149] 
07 ffff870e`66884840 fffff802`184579e0     OpenZFS!zfs_AcquireForLazyWrite+0x13e [H:\dev\openzfs\module\os\windows\zfs\zfs_vnops_windows.c @ 140] 
08 ffff870e`668848c0 fffff802`185057b1     nt!CcWriteBehindInternal+0x130
09 ffff870e`668849a0 fffff802`18502fe1     nt!CcWriteBehind+0x91
0a ffff870e`66884a90 fffff802`18450545     nt!CcCachemapUninitWorkerThread+0xf1
0b ffff870e`66884b70 fffff802`1850e6f5     nt!ExpWorkerThread+0x105
0c ffff870e`66884c10 fffff802`18606278     nt!PspSystemThreadStartup+0x55
0d ffff870e`66884c60 00000000`00000000     nt!KiStartSystemThread+0x28
```

This is here in `zfs_AcquireForLazyWrite`:

```c
	if (zfsvfs->z_unmounted ||
	    zfs_enter(zfsvfs, FTAG) != 0) {
```

When I check, `zfsvfs->z_unmounted` is already `TRUE`, so it was set *after* the mutex has been destroyed? I suspect this is also what crashed my GitHub actions runner in `tests.py` earlier even though it ran completely stable before.

---

`zfs_freevfs` and `zfsvfs_free` were called just before the crash. I think in `zfs_vfs_unmount` we need to set `zfsvfs->z_unmounted = B_TRUE;` before the `zfs_freevfs` at the end.

What is also curious is that in `zfs_AcquireForLazyWrite` `zfsvfs` is not NULL, but when checking `zmo->fsprivate` it is already zeroed, so the zeroing must have happened in between line 131 and 141.

---

Even if we moved `zfsvfs->z_unmounted = B_TRUE` in `zfs_vfs_unmount` we'd still potentially be accessing freed memory.

Another curious thing I found is:

```
FFFFA162A455D080: dprintf: zfs_vfsops.c:2062:zfs_freevfs(): +freevfs
FFFFA162A455D080: dprintf: zfs_vfsops.c:879:zfsvfs_free(): +zfsvfs_free
```

yet, there are no corresponding `-freevfs` / `-zfsvfs_free`, so it must have happened in the middle of freeing those.

---

It must have happened when `zfsvfs_free` was still before the `Unloading hardlink AVLtree` print

---

The unmount thread is in `zfsvfs_free` just after `ZFS_TEARDOWN_DESTROY(zfsvfs);`

---

To summarize: 

- `zfs_AcquireForLazyWrite` gets called with `vfs_fsprivate(zmo) != NULL`
- destruction of vfs is already running, sets `fsprivate=NULL`, frees mutex
- `zfs_AcquireForLazyWrite` continues execution but crashes because of freed mutex.

**Comment by lundman** [comment]
OK up to  260f628 attempts to handle this better, vfs_busy() is a rwlock, so now multiple acquire lazy/readahead/modwrite can run concurrently - thats probably a win, then in unmount, we make sure to grab a WRITER lock.

**Comment by EchterAgo** [comment]
I believe this can be closed, I haven't encountered any more issues during unmount with 260f628d650dcdfcf1323f877a66ac14b74536f5
